Define claims display description and claims path query (#276)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

> * `credential_definition`: REQUIRED. Object containing the detailed description of the Credential type. It consists of at least the following two parameters:

Reviewer: Isn't the sentence "It consists of at least the following two parameters" misleading because `credentialSubject` is marked as optional, so a valid `credential_definition` may also only contain one parameter?

I would suggest something like "It consists of at least the type parameter and an optional credentialSubject".

Author: This text is no longer present in the new version, so I'll close this.

  * `type`: REQUIRED. Array designating the types a certain Credential type supports, according to [@VC_DATA], Section 4.3.
  * `credentialSubject`: OPTIONAL. Object containing a list of name/value pairs, where each name identifies a claim offered in the Credential. The value can be another such object (nested data structures), or an array of such objects. To express the specifics about the claim, the most deeply nested value MAY be an object that includes the following parameters defined by this specification (other parameters MAY also be used):
      * `mandatory`: OPTIONAL. Boolean which, when set to `true`, indicates that the Credential Issuer will always include this claim in the issued Credential. If set to `false`, the claim is not included in the issued Credential if the wallet did not request the inclusion of the claim, and/or if the Credential Issuer chose to not include the claim. If the `mandatory` parameter is omitted, the default value is `false`.

>       * `value_type`: OPTIONAL. String value determining the type of value of the claim. Valid values defined by this specification are `string`, `number`, and image media types such as `image/jpeg` as defined in IANA media type registry for images (https://www.iana.org/assignments/media-types/media-types.xhtml#image). Other values MAY also be used.

Reviewer 2: The types `array` and `object` should also be mentioned explicitly and used in examples.
How can the type of array elements be determined?
```json
{"path": ["nationalities"], "value_type": "array"}
{"path": ["nationalities", null], "value_type": "string"}
```

Reviewer 3: #425

>   * `credentialSubject`: OPTIONAL. An array of claims description objects as defined in (#claims-description).

Reviewer 4: The use of an array of claim description objects now allows for multiple contradictory descriptions of the same claim, which I believe was not possible before.
E.g.
```
[
   {"path": ["nationalities"], "mandatory": true},
   {"path": ["nationalities"], "mandatory": false}
] 
```
We should clarify how a consumer should react to those cases:
- Reject the array (e.g. reject the authorization request is this happens on inside `authorization_details`)?
- Only consider the first one entry?
- Merge the properties of both entries (which may only make sense for some cases)?
With arrays, these collisions may be a bit more subtle.
E.g. How should this be handled on an authorization request?
```
[
   {"path": ["nationalities", 0], "mandatory": true},
   {"path": ["nationalities", null], "mandatory": false}
] 
```
Does this mean the `nationalities` array can have one or more elements.

Finally, we could have entries suggesting different structure (and not only different properties), such as
```
[
    {"path": ["nationalities", 0], "display": {...}},
    {"path": ["nationalities", "country"], "display": {...}}
]
```
The first entry indicates that `nationalities` is an array. The second entry indicates `nationalities` is an object.

Author: The first and second example show points we indeed need to address. The third one should be taken care of by the current processing rules already.

Author: I have fixed this in the latest revision.

* `credential_definition`: OPTIONAL. Object containing a detailed description of the Credential consisting of the following parameter:
  * `type`: OPTIONAL. Array as defined in (#server-metadata-jwt-vc-json). This claim contains the type values the Wallet requests authorization for at the Credential Issuer. It MUST be present if the claim `format` is present in the root of the authorization details object. It MUST not be present otherwise. 

>   * `credentialSubject`: OPTIONAL. An array of claims description objects as defined in (#claims-description), excluding the `display` and `value_type` parameters. The `mandatory` parameter here is used by the Wallet to indicate to the Issuer that it only accepts Credential(s) issued with those claim(s).

Reviewer 5: I don't think that referencing this way works very well given how the #claims-description and #claims-path-query sections are written currently.

>   * `credentialSubject`: OPTIONAL. An array of claims description objects as defined in (#claims-description), excluding the `display` and `value_type` parameters. The `mandatory` parameter here is used by the Wallet to indicate to the Issuer that it only accepts Credential(s) issued with those claim(s).

Reviewer 6: Could we start with the #claims-path-query section defining only `path` and then follow up with the #claims-description building on top?

>   * `credentialSubject`: OPTIONAL. An array of claims description objects as defined in (#claims-description), excluding the `display` and `value_type` parameters. The `mandatory` parameter here is used by the Wallet to indicate to the Issuer that it only accepts Credential(s) issued with those claim(s).

Reviewer 5: Probably, yeah. My comment here was just to say that this change (and others like it) doesn't work without changes to the referenced sections.

The following additional Credential Issuer metadata parameters are defined for this Credential Format for use in the `credential_configurations_supported` parameter, in addition to those defined in (#credential-issuer-parameters).

> * `claims`: OPTIONAL. An array of claims description objects as defined in (#claims-description-issuer-metadata).

Reviewer 7: given that `claims` is now generic across seemingly all credential formats can `claims` be added just once to the the Credential Type Metadata Parameters section?

> * `credential_definition`: REQUIRED. Object containing the detailed description of the Credential type. It consists of the following parameter:

Reviewer 3: why is `claims` not part of `credential_definition`?

Author: It is a top-level element for all other formats, so should be one here as well.

The following additional claims are defined for authorization details of type `openid_credential` and this Credential format.

* `doctype`: OPTIONAL. String as defined in (#server-metadata-mso-mdoc). This claim contains the type value the Wallet requests authorization for at the Credential Issuer. It MUST only be present if the `format` claim is present. It MUST not be present otherwise. 

> * `claims`: OPTIONAL. Object as defined in (#server-metadata-mso-mdoc), excluding the `display` and `value_type` parameters. The `mandatory` parameter here is used by the Wallet to indicate to the Issuer that it only accepts Credential(s) issued with those claim(s).

Reviewer 6: ```suggestion
* `claims`: OPTIONAL. Object containing a list of name/value pairs, where the name is a `namespace` as defined in [@!ISO.18013-5] (or any profile of it), and the value is an array of claims path queries as defined in (#claims-path-query). The  parameter is used by the Wallet to indicate to the Issuer that it only accepts Credential(s) issued containing exactly those claim(s). The namespace applies to the claims described in the array of claims description objects.
```

The value of the `credential` claim in the Credential Response MUST be a string that is an SD-JWT VC. Credentials of this format are already suitable for transfer and, therefore, they need not and MUST NOT be re-encoded.

# Claims Description {#claims-description}

> 

Reviewer 4: Currently the _claims description object_ described in this section is being used for both:
- the metadata (information produced by the issuer).
- `authorization_details` entries (RAR) and credential request (information produced by the wallet), where the `value_type` and `display` keys cannot be used.

Instead of having a single object model, I believe it would be preferable to have two distinct object model sections, both sharing the path definition. IMO, the advantages are:
- We can grow each object model independently. For instance, in the future we can add policy information to the metadata model, without needing to explicitly forbid those new elements on RAR and credential requests.
- Have better descriptions for each key. For instance, even if the `mandatory` key appears on both, we could have distinct fine-tuned descriptions for each case. Currently the `mandatory` description text only applies to its use on the metadata and the meaning of `mandatory` on RAR and credential requests needs to be inferred.

Reviewer 6: Hi Pedro, this is exactly what may suggestions do, please check

Author: Thanks, I have fixed this, please review the latest version.

A claims description object is an object used to describe how a certain claim in
the Credential should be displayed to the End-User. It is used in the `claims`
or `credentialSubject` parameters in the Credential Issuer metadata defined in
(#format-profiles).

For a claims description object, the following keys defined by this
specification MAY be used to describe the claim or claims (other keys MAY also
be used):

  * `path`: REQUIRED. Array. Claim path query as defined in (#claims-path-query)
    to identify the claim(s) in the Credential.
  * `mandatory`: OPTIONAL. Boolean which, when set to `true`, indicates that the
    Credential Issuer will always include this claim in the issued Credential.
    If set to `false`, the claim is not included in the issued Credential if the
    wallet did not request the inclusion of the claim, and/or if the Credential
    Issuer chose to not include the claim. If the `mandatory` parameter is
    omitted, the default value is `false`.

>   * `value_type`: OPTIONAL. String value determining the type of value of the

Reviewer 7: I feel that `value_type` needs more description. 
Is the default `string`?
The following value_types that are needed by our wallets are
1. iso8601_date
2. iso8601_time
3. iso8601_datetime
4. iso8601_duration
5. uri
6. image/* where the credential has a generic image type like the HTTP accepts header

Author: Please file this request for finer-grained value types as a separate issue.

Reviewer 3: #425

    claim. Valid values defined by this specification are `string`, `number`,
    and image media types such as `image/jpeg` as defined in IANA media type
    registry for images
    (https://www.iana.org/assignments/media-types/media-types.xhtml#image).
    Other values MAY also be used.
  * `display`: OPTIONAL. Array of objects, where each object contains display
    properties of a certain claim in the Credential for a certain language.
    Below is a non-exhaustive list of valid parameters that MAY be included:
     * `name`: OPTIONAL. String value of a display name for the claim.

>      * `locale`: OPTIONAL. String value that identifies language of this object

Reviewer 7: Open badge v3 credentials are used extensively throughout the education sector in the US. OBV3 supports an "identity" object that contains variable data. For example
```json
{
  "identityHash": "jane@example.com",
  "identityType": "email"
}
```
but the same structure can also hold a name
```json
{
  "identityHash": "Jane Foobar",
  "identityType": "name"
}
```

Something like the below would support it, but it requires conditional logic
```json
[{
  "path": ["identifier", "identityHash"],
  "display": [{
    "name": "Name",
    "path": ["identifier", "identityType"],
    "condition": "= 'name'"
  }, {
    "name": "Email",
    "path": ["identifier", "identityType"],
    "condition": "= 'emailAddress'"
  }]
}]
```
where condition must require `[< <= = != > >=] <literal>`.
It benefit being that it can be combined with the locales

Reviewer 6: Does this mean you have polymorphic claims?

Reviewer 7: @paulbastian yes, and it's not just me. https://1edtech.github.io/openbadges-specification/ob_v3p0.html#identityobject

Reviewer 6: I wouldn't support polymorphic claims, it seems this is a bad design pattern to me

Reviewer 7: I agree that it's not a good design. However, I believe that OpenBadgeCredential v3 is one of the top 2 most popular credential types in the world. So, deciding not to support this feature in OpenID4VCI needs to be done consciously and with outreach back to 1edtech

Reviewer 7: @paulbastian continued in https://github.com/openid/OpenID4VCI/issues/422

       represented as language tag values defined in BCP47 [@!RFC5646]. There
       MUST be only one object for each language identifier.

The order of claims description objects in the `claims` or `credentialSubject`
array is used by the Wallet to determine the order in which the claims are
displayed to the End-User, unless another mechanism is defined by the profile.

# Claims Path Query {#claims-path-query}

The metadata specifications above make use of claims path queries to select a
particular claim in the credential or a set of such claims. This appendix
defines the syntax and semantics of claims path queries.

> A claims path query MUST be a non-empty array of strings, `null` values, or

Reviewer 8: It would be nice to support negative integers as well since those are used by CWT/COSE claims.

Author: For addressing claims identified by numbers in CWT/COSE, I suggest to use numbers in strings to distinguish them from array elements. WDYT?

Reviewer 5: We don't yet have any Credential Format Profiles where claims identified by numbers in CWT/COSE.  But it might be good to have an approach for them anyway.  Using numbers in strings to distinguish them from array elements would work but then there's a potential (though unlikely) for ambiguity of a claim identified by the number vs. identified by the number as a string. Maybe the rules here could cover that rare case by just saying to select for both in the case that the credential format allows for it. 

Or use an object to represent a number claim name/key something like `{"claim-identified-by-number": -257}`

Reviewer 5: Another approach would be use numbers similar to strings as representing the literal key value to select for. And then a to use single element or empty arrays to convey selecting on array element(s). So something like `[1]` would be use where just `1` is being used now. And an empty `[]` array (or some variation) could be used in place of how `null` is being used now.

Something like this, for example: 

 `["degrees", [], "type"]`: All `type` claims in the `degrees` array are selected.
 `["nationalities", [1]]`: The second nationality is selected.

Reviewer 8: Another approach would be to use a different property instead of `path`. Perhaps the simplest option would be to use something like `path_cbor` that could then contain negative integer values?

Reviewer 9: I like this proposal, in particular using `[]` instead of `null` as a wildcard

I don't know it there might be issues with some implementations, as the ones mentioned for `""`. I can say that in my implementations I can handle void instanced of arrays or strings without any problem.

Reviewer 10: Using an Array would also allow you to
- select everything via `[]` 
- select a specific entry `[1]`
- select several sepcific entries (which other approaches can't easyil do `[0,1]`

That sounds like the cleanest option to me so far imho.

Author: We only have `null` in DCQL in OpenID4VP for now. I suggest that we extend this as need as soon as we have a concrete use case.

non-negative integers. 

> A string indicates that the respective key is to be selected, a `null` value

Reviewer 9: `null` -> `*`

would be more understandable ... but, before delving into these details, please consider the gap we have in IETF regarding the JSONPath (or similar approaches) and the requirement to get this standardized before propagating it in the openid specifications, in an unconventional way or in such a way that each specification, for specific needs, produces many narrow implementations with obvious implementation costs for the openid/ietf ecosystem.

As we know, JSONPath outlines a syntax for defining paths within a JSON document to find elements, much like XPath does for XML. It allows for filtering elements based on their attributes, navigating through arrays, and performing operations like slicing arrays.

Author: But `"*"` would be a string value which would require special rules for claims called `*` (admittedly, those shouldn't be too common).

JSONPath is a **huge** dependency. Suddenly, you have regular expressions and other stuff in your credential issuance protocol. Also some security issues, when you implement it incorrectly. And it brings so many things we don't need and would have to define-away...

Reviewer 9: I don't consider myself a jsonpath lover, I'm wondering if there's the time to define a basic, common and general purpose approach for dealing with all these json handling. I agree to not use JsonPath. I think that we have to define an alternative way, not just tied in openid4vci, something to be reused, and expandend where necessarly.

Considering my concerns about the use of null in claims path queries being potentially misleading and carrying a negative connotation, I'm wondering on an alternative approach with a more intuitive and self-explanatory keyword that aligns with the intended functionality. 

we have "special" string like "*" or "".

Or using objects like `{"select": "all"}` that unfortunately carries more nesting and other concerns.

Reviewer 10: I was thinking empty string as well instead of null (i think it's reasonable to not allow empty string names) but some libraries will probably also do weird stuff with an empty string.

<{{examples/credential_response_sd_jwt_vc.txt}}

# Claims Description 

Claims description objects are used in two places in this specification, in the
Credential Issuer metadata and in the Authorization Details. The following
sections define the structure and semantics of claims description objects.

## Claims Description for Authorization Details {#claims-description-authorization-details}

A claims description object as used in authorization details is an object that defines the
requirements for the claims that the Wallet requests to be included in the
Credential.

For a claims description object, the following keys defined by this
specification can be used to describe the claim or claims (other keys MAY also
be used):

  * `path`: REQUIRED. Array. Claim path pointer as defined in (#claims_path_pointer)
    to identify the claim(s) in the Credential.
  * `mandatory`: OPTIONAL. Boolean which, when set to `true`, indicates that the
    Wallet will only accept a Credential that includes this claim. If set to
    `false`, the claim is not required to be included in the Credential. If the

>     `mandatory` parameter is omitted, the default value is `false`.

Reviewer 3: good point. opened #433

A claims path pointer MUST be a non-empty array of strings, nulls and non-negative integers.
A claims path pointer can be processed, which means it is applied to a credential. The results of
processing are the referenced claims.

## Semantics for JSON-based credentials

This section defines the semantics of a claims path pointer when applied to a JSON-based credential.

A string value indicates that the respective key is to be selected, a null value
indicates that all elements of the currently selected array(s) are to be selected;
and a non-negative integer indicates that the respective index in an array is to be selected. The path
is formed as follows:

Start with an empty array and repeat the following until the full path is formed. 

  - To address a particular claim within an object, append the key (claim name)
    to the array.
  - To address an element within an array, append the index to the array (as a
    non-negative, 0-based integer).
  - To address all elements within an array, append a null value to the array.

### Processing

In detail, the array is processed from left to right as follows:

>  1. Select the root element of the Credential, i.e., the top-level JSON object.

Author: I'll change the wording to "Select the root element of the Credential, i.e., the top-level object of the
    JSON structure representing the claims in the Credential"

<{{examples/credential_response_sd_jwt_vc.txt}}

# Claims Description 

Claims description objects are used in two places in this specification, in the
Credential Issuer metadata and in the Authorization Details. The following
sections define the structure and semantics of claims description objects.

## Claims Description for Authorization Details {#claims-description-authorization-details}

A claims description object as used in authorization details is an object that defines the
requirements for the claims that the Wallet requests to be included in the
Credential.

For a claims description object, the following keys defined by this
specification can be used to describe the claim or claims (other keys MAY also
be used):

  * `path`: REQUIRED. Array. Claim path pointer as defined in (#claims_path_pointer)
    to identify the claim(s) in the Credential.
  * `mandatory`: OPTIONAL. Boolean which, when set to `true`, indicates that the
    Wallet will only accept a Credential that includes this claim. If set to
    `false`, the claim is not required to be included in the Credential. If the

>     `mandatory` parameter is omitted, the default value is `false`.

Reviewer 6: I think authorization details really starts making sense, if we add `value` here, letting the Client say that he wants a Credential with particular values, which would be possible now. Is there any support for this?

Reviewer 10: relevant issue: #262
I still think it would make sense to add as an optional feature for cases where you already know what you want (e.g., from another app) and want to provide a hint to the Issuer.

Reviewer 11: Discussed on the WG call today: tendency to do the change and do it now in VP and VCI.

Reviewer 3: Please open another issuer/PR so that we can discuss it there

Reviewer 3: #426 (feel free to rename, not sure i got the name right)

The rules defined in (##claims-description-processing) apply.

## Claims Description for Issuer Metadata {#claims-description-issuer-metadata}

A claims description object as used in the Credential Issuer metadata is an
object used to describe how a certain claim in the Credential should be
displayed to the End-User. It is used in the `claims`
parameter in the Credential Issuer metadata defined in (#format-profiles). The
following keys can be used to describe the claim or claims:

  * `path`: REQUIRED if the Credential Format uses a JSON-based claims

>     structure; MUST NOT be present otherwise. The value MUST be a non-empty

Reviewer 12: I think this is discriminatory to Credential Formats using YAML based claims structures... ;-)

More seriously I do wonder if this goes a little too far with the normative language - it seems unnecessary to completely prevent someone defining a new non-JSON hierarchical credential format from using the `path` key for example.

Reviewer 3: will be resolved with the below suggestion

>   * `path`: REQUIRED if the Credential Format uses a JSON-based claims
>     structure; MUST NOT be present otherwise. The value MUST be a non-empty
>     array representing a claims path pointer that specifies the path to a claim
>     within the credential, as defined in (#claims_path_pointer).
>   * `namespace`: REQUIRED if the Credential Format is based on the mdoc format
>     defined in ISO 18013-5; MUST NOT be present otherwise. The value MUST be a
>     string that specifies the namespace of the data element within the mdoc,
>     e.g., `org.iso.18013.5.1`.

Reviewer 3: ```suggestion
  * `path`: REQUIRED if the Credential Format uses a JSON-based claims
    structure; MUST NOT be present otherwise. The value MUST be a non-empty
    array representing a claims path pointer that specifies the path to a claim
    within the credential, as defined in (#claims_path_pointer).
  * `namespace`: REQUIRED if the Credential Format is based on the mdoc format
    defined in ISO 18013-5; MUST NOT be present otherwise. The value MUST be a
    string that specifies the namespace of the data element within the mdoc,
    e.g., `org.iso.18013.5.1`.
```

>   * `path`: REQUIRED if the Credential Format uses a JSON-based claims
>     structure; MUST NOT be present otherwise. The value MUST be a non-empty
>     array representing a claims path pointer that specifies the path to a claim
>     within the credential, as defined in (#claims_path_pointer).
>   * `namespace`: REQUIRED if the Credential Format is based on the mdoc format
>     defined in ISO 18013-5; MUST NOT be present otherwise. The value MUST be a
>     string that specifies the namespace of the data element within the mdoc,
>     e.g., `org.iso.18013.5.1`.
>   * `claim_name`: REQUIRED if the Credential Format is based on mdoc format
>     defined in ISO 18013-5; MUST NOT be present otherwise. The value MUST be a
>     string that specifies the data element identifier of the data element within
>     the provided namespace in the mdoc, e.g., `first_name`.

Reviewer 3: ```suggestion
  * `path`: REQUIRED if the Credential Format uses a JSON-based claims
    structure; SHOULD NOT be present otherwise. The value MUST be a non-empty
    array representing a claims path pointer that specifies the path to a claim
    within the credential, as defined in (#claims_path_pointer).
  * `namespace`: REQUIRED if the Credential Format is based on the mdoc format
    defined in ISO 18013-5; SHOULD NOT be present otherwise. The value MUST be a
    string that specifies the namespace of the data element within the mdoc,
    e.g., `org.iso.18013.5.1`.
  * `claim_name`: REQUIRED if the Credential Format is based on mdoc format
    defined in ISO 18013-5; SHOULD NOT be present otherwise. The value MUST be a
    string that specifies the data element identifier of the data element within
    the provided namespace in the mdoc, e.g., `first_name`.
```

>     the provided namespace in the mdoc, e.g., `first_name`.

Reviewer 12: I wonder if we should change (remove namespace/claim_name) as per the discussion on https://github.com/openid/OpenID4VP/issues/293

Reviewer 3: +1 to changing

  * `mandatory`: OPTIONAL. Boolean which, when set to `true`, indicates that the
    Credential Issuer will always include this claim in the issued Credential.
    If set to `false`, the claim is not included in the issued Credential if the
    wallet did not request the inclusion of the claim, and/or if the Credential
    Issuer chose to not include the claim. If the `mandatory` parameter is
    omitted, the default value is `false`.

>   * `value_type`: OPTIONAL. String value determining the type of value of the

Reviewer 6: To me the value_type doesn't seem very thoughtthrough/complete, maybe its better to leave it out for this merge and add it later if we really want it

Reviewer 7: @paulbastian i agree (I made a comment earlier listing a number of `value_type` values that are required to correctly format rendered claims). I think removing `value_type` for now and adding it back in with more consideration would make sense

Reviewer 13: @paulbastian , @sloops77 I also agree with this (to remove `value_type` from this PR)

Current values (`string`, `number` etc) make little sense for `mdoc`, as  @sloops77 commented a while ago.

I tend to believe that this is also the case for JSON based credentials. 
For instance, is it helpful to have for claim `birthdate` a value_type `string`?

    claim. Valid values defined by this specification are `string`, `number`,

>     and image media types such as `image/jpeg` as defined in IANA media type

Reviewer 13: `boolean` ?

Reviewer 3: removing value_type for now

 - there is a claims description object indicating that a certain claim is an array
   (implied by using `null` or a non-negative integer in the `path`), and another
   object indicating that the same claim is an object (implied by using a string
   in the `path`).

# Claims Path Pointer {#claims_path_pointer}

A claims path pointer is a pointer into the JSON structure of the Verifiable
Credential, identifying one or more claims. A claims path pointer MUST be a
non-empty array of strings and non-negative integers. A string value 
indicates that the respective key is to be selected, a null value
indicates that all elements of the currently selected array(s) are to be selected;
and a non-negative integer indicates that the respective index in an array is to be selected. The path
is formed as follows:

Start with an empty array and repeat the following until the full path is formed. 

  - To address a particular claim within an object, append the key (claim name)
    to the array.
  - To address an element within an array, append the index to the array (as a
    non-negative, 0-based integer).
  - To address all elements within an array, append a null value to the array.

> Verifiers MUST NOT point to the same claim more than once in a single query.

Reviewer 7: i think this should be removed or rephrased as Verifiers are not relevant to creating claims descriptions

Reviewer 3: verifiers need to be changed to Credential issuers.
